File: lib/api/query.ts
import {
  getStreamApiBase,
  ApiError,
  getStoredToken,
  clearStoredToken,
} from "@/lib/api/client";
import { normalizeSources } from "@/lib/api/normalize";
import type { QuerySource } from "@/lib/types";

export type QueryMode =
  | "rag"
  | "doc_relations"
  | "list_documents"
  | "document_summary"
  | "general_chat"
  | "knowledge_qa";

/** Query Router 的判定结果（架构图 Query Router 节点）。 */
export interface RouteInfo {
  intent: string;
  reason: string;
}

/** Retrieval Grader 的判定结果（架构图 Retrieval Grader 节点）。 */
export interface GradeInfo {
  good: boolean;
  retry: number;
  reason: string;
}

export interface StreamQueryOptions {
  query: string;
  collectionId?: string | null;
  conversationId?: string | null;
  /** "doc_relations" enables the cross-document relation pipeline (问题1). */
  mode?: QueryMode;
  signal?: AbortSignal;
  onToken: (token: string) => void;
  onThinking?: (delta: string) => void;
  onSources: (sources: QuerySource[]) => void;
  onDone: (conversationId?: string) => void;
  onError: (error: Error) => void;
  /** 可选：Query Router 判定结果（用于展示"识别为文档总结/闲聊"等状态）。 */
  onRoute?: (info: RouteInfo) => void;
  /** 可选：Retrieval Grader 判定结果（用于展示"证据不足，正在重试"）。 */
  onGrade?: (info: GradeInfo) => void;
}

/** Map a doc_digests SSE event entry to the shared QuerySource shape. */
function normalizeDigest(raw: Record<string, unknown>): QuerySource {
  const pageCount = Number(raw.page_count ?? 0);
  return {
    documentId: raw.document_id ? String(raw.document_id) : undefined,
    documentName: raw.filename ? String(raw.filename) : "未知文档",
    chunkText: typeof raw.digest === "string" ? raw.digest : undefined,
    pages: pageCount > 0 ? pageCount : undefined,
  };
}

type SseHandler = Pick<
  StreamQueryOptions,
  "onToken" | "onThinking" | "onSources" | "onDone" | "onError" | "onRoute" | "onGrade"
>;

/** 把单个 SSE data 行分发到对应回调。 */
function dispatchSseData(data: string, handlers: SseHandler) {
  if (data === "[DONE]") {
    handlers.onDone();
    return;
  }
  try {
    const parsed = JSON.parse(data) as Record<string, unknown>;
    const type = String(parsed.type ?? parsed.event ?? "");

    // Query Router 判定结果（新增事件；旧后端不会发，忽略即可）
    if (type === "intent") {
      handlers.onRoute?.({
        intent: String(parsed.intent ?? ""),
        reason: String(parsed.reason ?? ""),
      });
      return;
    }
    // Retrieval Grader 判定结果：good=false 表示正在改写重试
    if (type === "grade") {
      handlers.onGrade?.({
        good: parsed.good === true,
        retry: Number(parsed.retry ?? 0),
        reason: String(parsed.reason ?? ""),
      });
      return;
    }
    if (type === "sources" || parsed.sources) {
      handlers.onSources(normalizeSources(parsed.sources ?? parsed.data));
      return;
    }
    // Cross-document relation analysis (问题1): per-document digests
    if (type === "doc_digests") {
      const docs = Array.isArray(parsed.documents) ? parsed.documents : [];
      handlers.onSources(
        (docs as Record<string, unknown>[]).map(normalizeDigest)
      );
      return;
    }
    if (type === "thinking_delta") {
      handlers.onThinking?.(String(parsed.content ?? ""));
      return;
    }
    if (type === "error") {
      handlers.onError(
        new Error(String(parsed.message ?? parsed.content ?? "生成失败"))
      );
      return;
    }
    if (type === "done") {
      handlers.onDone(
        parsed.conversation_id ? String(parsed.conversation_id) : undefined
      );
      return;
    }

    const token = String(
      parsed.content ??
        parsed.token ??
        parsed.text ??
        parsed.delta ??
        (type === "token" ? parsed.data : "") ??
        ""
    );
    if (token) handlers.onToken(token);
  } catch {
    if (data) handlers.onToken(data);
  }
}

export async function streamQuery(options: StreamQueryOptions): Promise<void> {
  // 直连 FastAPI 后端（绕过 Next.js 代理），确保 SSE 逐 token 到达
  const base = getStreamApiBase();
  const token = getStoredToken();

  let res: Response;
  try {
    res = await fetch(`${base}/query`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        Accept: "text/event-stream",
        ...(token ? { Authorization: `Bearer ${token}` } : {}),
      },
      body: JSON.stringify({
        query: options.query,
        collection_id: options.collectionId ?? undefined,
        conversation_id: options.conversationId ?? undefined,
        mode: options.mode ?? undefined,
        stream: true,
      }),
      signal: options.signal,
    });
  } catch (err) {
    if ((err as Error).name === "AbortError") throw err;
    throw new ApiError(0, "无法连接服务器，请确认后端服务已启动");
  }

  if (res.status === 401) {
    clearStoredToken();
    if (typeof window !== "undefined" && window.location.pathname !== "/login") {
      window.location.href = "/login";
    }
    throw new ApiError(401, "登录已过期，请重新登录");
  }
  if (res.status === 429) {
    throw new ApiError(429, "提问太频繁了，请稍后再试");
  }
  if (!res.ok) {
    const text = await res.text().catch(() => "");
    throw new ApiError(res.status, text || res.statusText || "请求失败");
  }

  const contentType = res.headers.get("content-type") ?? "";

  // 非流式回退：后端直接返回 JSON（理论上不会发生，保留兜底）
  if (!contentType.includes("text/event-stream")) {
    const json = (await res.json()) as Record<string, unknown>;
    if (json.sources) {
      options.onSources(normalizeSources(json.sources));
    }
    const answer = String(json.answer ?? json.response ?? json.content ?? "");
    if (answer) options.onToken(answer);
    options.onDone();
    return;
  }

  if (!res.body) throw new ApiError(500, "浏览器不支持流式读取");

  // ── 流式主路径：逐行解析 SSE ─────────────────────────────────────────────
  // 不依赖事件间空行的成组切分，任何一行到达立即处理 —— token 即到即渲染。
  const reader = res.body.getReader();
  const decoder = new TextDecoder();
  let buffer = "";
  let finished = false;
  const onDoneOnce = (conversationId?: string) => {
    if (finished) return;
    finished = true;
    options.onDone(conversationId);
  };
  const handlers: SseHandler = { ...options, onDone: onDoneOnce };

  const handleLine = (rawLine: string) => {
    const line = rawLine.replace(/\r$/, "");
    if (!line.startsWith("data:")) return;
    const data = line.slice(5).trim();
    if (!data) return;
    dispatchSseData(data, handlers);
  };

  try {
    while (true) {
      const { done, value } = await reader.read();
      if (done) break;
      buffer += decoder.decode(value, { stream: true });
      let newlineIdx: number;
      while ((newlineIdx = buffer.indexOf("\n")) !== -1) {
        const line = buffer.slice(0, newlineIdx);
        buffer = buffer.slice(newlineIdx + 1);
        handleLine(line);
      }
    }
    if (buffer.trim()) handleLine(buffer.trim());
  } catch (err) {
    if ((err as Error).name === "AbortError") {
      return; // 用户主动停止
    }
    throw err;
  }

  onDoneOnce();
}

File: backend/app/api/query.py
"""
RAG query API router (Phase 4 / 问题1).

POST /query — accepts a JSON body, runs the LangGraph RAG pipeline,
              and streams the response as Server-Sent Events (SSE).

Two pipelines (selected by ``mode``, auto-detected when omitted):
    rag            — semantic retrieval QA (default)
    doc_relations  — cross-document relation analysis (问题1): answers the
                     "这些文档有什么关联" question organised BY DOCUMENT

SSE event stream format
────────────────────────
Every event is a JSON object on a ``data:`` line, terminated by
two newlines (per the SSE spec).

    data: {"type":"thinking"}

    data: {"type":"sources","sources":[...],"retrieved_count":5}
      — or, for doc_relations mode:
    data: {"type":"doc_digests","documents":[...],"retrieved_count":5}

    data: {"type":"chunk","content":"The annual report..."}

    data: {"type":"chunk","content":" shows revenue growth..."}

    data: {"type":"done","conversation_id":"<uuid>","total_chars":412}

    data: [DONE]

Error event (only when the pipeline fails mid-stream):

    data: {"type":"error","message":"<description>"}

    data: [DONE]

Client-side usage example (JavaScript):

    const es = new EventSource(undefined);
    const resp = await fetch("/query", {
      method: "POST",
      headers: {"Content-Type": "application/json"},
      body: JSON.stringify({query: "...", conversation_id: null}),
    });
    const reader = resp.body.getReader();
    // read SSE lines and parse JSON from each `data:` line
"""

import json
import uuid
from typing import Annotated, AsyncGenerator

from fastapi import APIRouter, Depends, HTTPException, Request, status
from fastapi.responses import StreamingResponse

from app.api.deps import client_ip, enforce_rate_limit
from app.services.permissions import require_permission
from app.config import get_settings
from app.db.user_models import User
from app.schemas.query import QueryRequest, SSEThinking
from app.services.audit_service import record_audit
from app.services.conversation_service import get_or_create_conversation, load_history
from app.services.prompt_security import inspect_user_query
from app.services.master_graph import stream_master
from app.services.rag_graph import stream_document_list
from app.services.routers.intent_rules import deterministic_route
from app.utils.logging import get_logger

logger = get_logger(__name__)

router = APIRouter(tags=["Query"])

# SSE helpers ──────────────────────────────────────────────────────────────────

def _sse(payload: dict | str) -> str:
    """Format a single SSE data line."""
    if isinstance(payload, str):
        return f"data: {payload}\n\n"
    return f"data: {json.dumps(payload, ensure_ascii=False)}\n\n"


# Deterministic pipeline auto-detection (问题1) ──────────────────────────────
#
# 规则本体在 app/services/routers/intent_rules.py，与 master graph 的
# Query Router 共用同一份定义，避免两边正则漂移。
#
# 注意返回值语义（问题1 的关键）：判定不了时返回 **None**，而不是以前的
# "rag"。返回 "rag" 会让调用方误以为"已确定走知识库问答"，进而绕过 LLM
# 路由 —— 这正是闲聊分支永远触发不了的原因之一。


def _detect_mode(query: str) -> str | None:
    """Return the deterministically selected pipeline, or None if undecided."""
    detected = deterministic_route(query)
    if detected in (
        "doc_relations",
        "list_documents",
        "general_chat",
        "document_agent",
    ):
        logger.info("Query auto-detected as %s: %r", detected, query[:80])
        return detected
    return None


# 旧 mode 名 → master graph intent 名。
# 自动检测（未显式传 mode）时不走这里，交给 master graph 的 LLM router，
# 它比正则更准；只有客户端显式指定 mode 才需要归一化。
_LEGACY_MODE_ALIASES = {
    "rag": "knowledge_qa",
}


# Streaming generator ──────────────────────────────────────────────────────────

async def _generate_sse(
    request: Request,
    query: str,
    conversation_id: uuid.UUID | None,
    top_k: int,
    mode: str,
    user: User,
    collection_id: uuid.UUID | None = None,
    forced_mode: str | None = None,
) -> AsyncGenerator[str, None]:
    """
    Async generator that drives the entire RAG pipeline and yields raw SSE
    strings suitable for a StreamingResponse.

    Pipeline:
        1. Resolve / create conversation (PostgreSQL, owner-scoped)
        2. Load conversation history (PostgreSQL)
        3. Yield "thinking" event immediately
        4. Run the master LangGraph astream_events:
             a. After the router node   → yield "intent"   (路由结果)
             b. After retrieval/digests → yield "sources" / "doc_digests"
             c. After the grader        → yield "grade"    (good / bad + retry)
             d. Per Ollama token        → yield "chunk"    (1:1 与模型输出)
             e. After output_guard      → yield "output_guard" (问题3+4 合规审计信号)
             f. After graph end         → yield "done"
        5. Yield "[DONE]" SSE terminator

    *mode* is the auto-detected/legacy pipeline label used for logging and for
    the deterministic ``list_documents`` fast path. *forced_mode* is only set
    when the client explicitly passed ``mode`` in the request body — in that
    case the master graph's LLM router is bypassed.

    Retrieval is restricted to documents owned by *user* (admins see all)
    and, when *collection_id* is set, to one knowledge-base collection.
    """
    # ── 1+2. Conversation setup (owner + tenant scoped) ───────────────────────
    from app.services.tenancy import scope_for

    # 可见范围一处组装：平台管理员 = 全平台（跨公司部门库/公司库，但看不到
    # 别人的个人库）；其他人 = 本公司内 个人 + 本部门 + 公司库。
    scope = scope_for(user)
    # 个人库归属**恒为本人**（含平台管理员）：owner_id 不再是"None = admin 全览"，
    # 因此这里必须是 user.id，否则管理员会连自己的个人库都检索不到。
    owner = scope.owner_id
    tenant = scope.tenant_id
    department = scope.department_id
    try:
        conv_id = await get_or_create_conversation(
            conversation_id, owner_id=owner, tenant_id=tenant,
        )
        history = await load_history(conv_id, owner_id=owner, tenant_id=tenant)
    except Exception as exc:
        logger.exception("Conversation setup failed: %s", exc)
        yield _sse({"type": "error", "message": f"Conversation setup failed: {exc}"})
        yield _sse("[DONE]")
        return

    # ── 3. Immediate "thinking" event ─────────────────────────────────────────
    yield _sse(SSEThinking().model_dump())

    # ── 4. LangGraph streaming ────────────────────────────────────────────────
    if mode == "list_documents":
        # 文档列表走确定性 DB 直读 —— 不进 LLM、不检索，保证不漏不重
        graph_events = stream_document_list(
            query=query,
            conversation_id=str(conv_id),
            owner_id=str(owner) if owner else None,
            collection_id=str(collection_id) if collection_id else None,
            tenant_id=tenant,
            user_department_id=department,
            user_id=str(user.id),
            tenant_wide=scope.tenant_wide,
            platform_wide=scope.platform_wide,
        )
    else:
        # 其余全部交给 master graph：
        #   Query Router → {Document Summary | Knowledge QA(Hybrid RAG
        #   + Retrieval Grader + Retry) | General Chat | doc_relations}
        graph_events = stream_master(
            query=query,
            conversation_id=str(conv_id),
            history_messages=history,
            top_k=top_k,
            owner_id=str(owner) if owner else None,
            collection_id=str(collection_id) if collection_id else None,
            forced_mode=forced_mode,
            # 监控 / Bad Case 回流需要"是谁问的"，owner_id 在管理员视角下为 None，
            # 因此单独传真实用户身份（与检索权限解耦）。
            user_id=str(user.id),
            username=user.username,
            # 三层隔离：第一层公司前置过滤 + 第二层部门 ACL + 第三层会话归属。
            # tenant_id=None 仅表示"跨公司"（平台管理员），仍受 ACL 约束。
            tenant_id=tenant,
            department_id=department,
            tenant_wide=scope.tenant_wide,
            platform_wide=scope.platform_wide,
        )

    try:
        async for graph_event in graph_events:
            # Check for client disconnect on each event to stop early
            if await request.is_disconnected():
                logger.info("Client disconnected mid-stream for conv=%s", conv_id)
                return

            yield _sse(graph_event)

    except Exception as exc:
        logger.exception("RAG graph stream error for conv=%s: %s", conv_id, exc)
        yield _sse({"type": "error", "message": str(exc)})

    # ── 5. SSE stream terminator ──────────────────────────────────────────────
    yield _sse("[DONE]")


# Router endpoint ──────────────────────────────────────────────────────────────

@router.post(
    "/query",
    summary="RAG query with streaming response",
    description=(
        "Submit a natural-language question. The response is a "
        "**Server-Sent Events** stream.\n\n"
        "Set `Accept: text/event-stream` or simply consume the stream — "
        "FastAPI will set `Content-Type: text/event-stream` automatically.\n\n"
        "**Query Router**: when `mode` is omitted, an LLM router classifies the "
        "question and dispatches it to one of five pipelines — `knowledge_qa` "
        "(Hybrid RAG: 多路向量 + BM25 → RRF → cross-encoder 精排), "
        "`document_summary` (per-document content summary), `general_chat` "
        "(no retrieval at all), `doc_relations` (cross-document analysis) and "
        "`list_documents` (deterministic DB listing). Passing `mode` explicitly "
        "bypasses the router.\n\n"
        "**Retrieval Grader + Retry**: in `knowledge_qa`, retrieved evidence is "
        "graded by an LLM (not just a score threshold). If the evidence cannot "
        "answer the question, the query is rewritten and retrieval retried up to "
        "`RETRIEVAL_MAX_RETRIES` times before refusing to answer.\n\n"
        "**Event types** (in order):\n"
        "- `thinking` — pipeline started\n"
        "- `intent`   — router decision (`intent`, `reason`)\n"
        "- `sources` / `doc_digests` — retrieved metadata (before any tokens)\n"
        "- `grade`    — evidence grader result (`good`, `retry`, `reason`)\n"
        "- `chunk`    — one LLM token or small token batch\n"
        "- `done`     — answer complete; includes `conversation_id` for follow-up\n"
        "- `error`    — unrecoverable failure (stream still closes cleanly)\n\n"
        "Pass the returned `conversation_id` in subsequent requests to continue "
        "the conversation with full memory."
    ),
    response_class=StreamingResponse,
    responses={
        200: {
            "description": "SSE stream of RAG events",
            "content": {"text/event-stream": {}},
        },
        422: {"description": "Validation error — invalid request body"},
    },
)
async def query_endpoint(
    body: QueryRequest,
    request: Request,
    user: Annotated[User, Depends(require_permission("conversation.write"))],
) -> StreamingResponse:
    settings = get_settings()

    # Per-user rate limit — protects the shared Ollama instance
    await enforce_rate_limit(
        request,
        scope="query",
        key=str(user.id),
        limit=settings.QUERY_RATE_LIMIT,
        window_seconds=settings.RATE_LIMIT_WINDOW_SECONDS,
    )

    # Prompt-injection guard runs before query rewrite, retrieval and LLM use.
    # Blocked input never reaches the vector store or model; suspicious-but-
    # legitimate security questions are allowed and marked in the audit trail.
    inspection = inspect_user_query(body.query)
    effective_query = inspection.normalized_text
    if settings.PROMPT_GUARD_ENABLED and inspection.blocked and settings.PROMPT_GUARD_BLOCK_HIGH_RISK:
        await record_audit(
            "security.prompt_injection.blocked",
            user_id=user.id,
            username=user.username,
            resource_type="query",
            resource_id=str(body.collection_id) if body.collection_id else None,
            detail=f"risk={inspection.risk}; signals={len(inspection.reasons)}; q={effective_query[:160]}",
            ip=client_ip(request),
        )
        logger.warning(
            "Blocked suspected prompt injection user=%s signals=%d",
            user.username,
            len(inspection.reasons),
        )
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="该请求包含疑似试图改变系统行为的指令，已被安全策略拦截。请仅提交与知识库内容相关的问题。",
        )

    detected = _detect_mode(effective_query)
    # "auto" 表示交给 master graph 的 LLM 路由决定（不再是以前的 "rag"）。
    mode = body.mode or detected or "auto"

    # 客户端显式指定 mode 时，跳过 master graph 的 LLM 路由。
    # 旧客户端习惯传 mode="rag"，这里归一化成 master graph 的 intent 名。
    forced_mode = None
    if body.mode:
        forced_mode = _LEGACY_MODE_ALIASES.get(body.mode, body.mode)
    elif detected in ("doc_relations", "general_chat", "document_agent"):
        # 确定性判定命中 → 跳过 LLM 路由，省一次调用，且不受路由超时影响。
        # list_documents 不走这里，它有自己的 DB 直读快路径。
        forced_mode = detected

    await record_audit(
        "query.ask" if inspection.risk == "clean" else "query.ask.suspicious",
        user_id=user.id,
        username=user.username,
        resource_type="query",
        resource_id=str(body.collection_id) if body.collection_id else None,
        detail=f"mode={mode}; risk={inspection.risk}; q={effective_query[:300]}",
        ip=client_ip(request),
    )

    logger.info(
        "POST /query | conv=%s top_k=%d mode=%s user=%s collection=%s risk=%s query=%r",
        body.conversation_id,
        body.top_k,
        mode,
        user.username,
        body.collection_id,
        inspection.risk,
        effective_query[:80],
    )

    return StreamingResponse(
        _generate_sse(
            request=request,
            query=effective_query,
            conversation_id=body.conversation_id,
            top_k=body.top_k,
            mode=mode,
            user=user,
            collection_id=body.collection_id,
            forced_mode=forced_mode,
        ),
        media_type="text/event-stream",
        headers={
            "Cache-Control": "no-cache",
            "Connection": "keep-alive",
            "X-Accel-Buffering": "no",       # disable nginx proxy buffering
            "Access-Control-Allow-Origin": "*",
        },
    )
